### PR TITLE
Dependency: Upgrade Yarn to v4

### LINKED ISSUE #24552: Investigate upgrading monorepo to Yarn 4
```[tasklist]
### Tasks
- [ ] Does yarn pnp still work? We are using @yarnpkg/fslib to support pnp
```

## Pnp Support

As of https://github.com/yarnpkg/berry/blob/master/CHANGELOG.md#api-changes, there are changes regarding the pnp API. In the past, we already tried out the release candidate v4.x of @yarnpkg/libzip, which dropped Node v.16 support. This was the reason to roll back to v3. We should verify whether [our implementation](https://github.com/storybookjs/storybook/blob/next/code/lib/cli/src/js-package-manager/Yarn2Proxy.ts#L130) of pnp support still works with yarn v4.

### PR DESCRIPTION
**Pull request by JReinhold**
Fixes #24552 

## What I did

Upgrade our internal usage of Yarn to v4.

References:

- https://yarnpkg.com/blog/release/4.0
- https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F4.0.0
- https://github.com/yarnpkg/berry/issues/3591

#### The changes in this PR are covered in the following automated tests:
- [x] stories
- [x] unit tests
- [x] integration tests
- [x] end-to-end tests

#### Manual testing

1. Checkout the branch
2. Run `git clean -xdf` (beware, will delete all uncommitted files)
3. Run `yarn start` which should take you through the whole tasks pipeline and see that it works
4. `git clean -xdf` again
5. `yarn task --task=compile --no-link` to see that it also compiles in `--no-link` mode
6. `cd scripts`
7. `yarn release:version --exact 0.0.0-canary-test-yarn.0`
8. `yarn release:publish --tag canary` to see that it correctly attempts to publish. It should output a long list of `YN0033: No authentication configured for request` errors because you don't have the npm token set, which is the correct output.

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24565-sha-2d556c14`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24565-sha-2d556c14). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24565-sha-2d556c14`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24565-sha-2d556c14) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`upgrade-yarn-4`](https://github.com/storybookjs/storybook/tree/upgrade-yarn-4) |
| **Commit** | [`2d556c14`](https://github.com/storybookjs/storybook/commit/2d556c1426bab010edad083bc473f22c68917fcf) |
| **Datetime** | Tue Oct 24 11:57:48 UTC 2023 (`1698148668`) |
| **Workflow run** | [6626437775](https://github.com/storybookjs/storybook/actions/runs/6626437775) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24565`_
</details>

### CODE REVIEW

File: scripts/release/publish.ts [hunk 1]
```diff
@@ -124,7 +124,7 @@ const publishAllPackages = async ({
   dryRun?: boolean;
 }) => {
   console.log(`📦 Publishing all packages...`);
-  const command = `yarn workspaces foreach --parallel --no-private --verbose npm publish --tolerate-republish --tag ${tag}`;
+  const command = `yarn workspaces foreach --all --parallel --no-private --verbose npm publish --tolerate-republish --tag ${tag}`;
```
Comment (valentinpalkovic) [inline scripts/release/publish.ts:127]: The previous default in v3 was `worktree` instead of `all`. Can you explain, why we use `all` now?

Comment (JReinhold) [reply scripts/release/publish.ts:127]: TBH I didn't understand the difference between "worktree" and "all", especially in our scenario, so I just went with what they have in the example in the docs.

File: .circleci/config.yml [hunk 2]
```diff
@@ -84,23 +84,23 @@ commands:
 jobs:
   pretty-docs:
     executor:
-      class: small
+      class: medium
```
Comment (JReinhold) [inline .circleci/config.yml:87]: CI would consistently silently bail on `yarn install`, this fixed that. We were likely running out of memory.

See "Install" step here silently stopping at "Fetch" step, causing the command to fail completely later.

https://app.circleci.com/pipelines/github/storybookjs/storybook/61775/workflows/4e8ba21c-8f6e-4d2b-b08d-46a6b0c764ab/jobs/588725

File: .yarn/plugins/@yarnpkg/plugin-workspace-tools.cjs
Comment (JReinhold) [inline .yarn/plugins/@yarnpkg/plugin-workspace-tools.cjs:1]: autodeleted by Yarn 4, included by default in core now

File: .yarnrc.yml [hunk 3]
```diff
@@ -1,11 +1,11 @@
+compressionLevel: mixed
+
+enableGlobalCache: false
```
Comment (JReinhold) [inline .yarnrc.yml:3]: this was auto set by Yarn 4. I believe this was the default in Yarn 3 but not Yarn 4, so I didn't want to change it.

File: scripts/release/publish.ts (hunk 1 above)
Comment (valentinpalkovic) [reply scripts/release/publish.ts:127]: I guess that it is context-specific. If `process.cwd()` would return a subfolder, yarn would only consider all projects in that subfolder when `--worktree` is used. With `--all` yarn will consider all projects defined in the root package.json

**Comment by valentinpalkovic** [comment]
@JReinhold We should definitely run a snapshot release on this branch to see whether the `workspace:*` versions get properly replaced.

File: scripts/release/publish.ts (hunk 1 above)
Comment (JReinhold) [reply scripts/release/publish.ts:127]: I see that makes sense, so it doesn't make a difference for us.

**Comment by yannbf** [comment]
LGTM! I'd recommend doing what Valentin said, other than that it looks good.

**Comment by JReinhold** [comment]
@valentinpalkovic @yannbf I've released a canary and tested it out in Mealdrop, it works fine.

**Review by valentinpalkovic — APPROVED** [review]

**Pull request merged**